### PR TITLE
Improving S3 File System

### PR DESCRIPTION
**Pull request by rahul003**
**Usability:**
- Adds an environment variable TF_S3_LOG_LEVEL to control the logging information from S3 operations. 
  - By default, this level is set to 1, which only logs fatal errors. This removes the flood of unnecessary logs that are thrown when using S3 currently which stem from improper usage of AWS_CPP_SDK logging levels. 
  - User can change the level when necessary using the following values: 0 (OFF), 1 (FATAL), 2 (ERROR), 3 (WARNING), 4 (INFO), 5 (DEBUG), 6 (TRACE)

**Reliability**
S3 supports an [eventual consistency model](https://docs.aws.amazon.com/AmazonS3/latest/dev/Introduction.html#ConsistencyModel), which does not work like a regular file system. This means that many operations currently in the s3 file system crash now and then because the conditions it expects are not satisfied by S3. The following changes tries to address such crashes. 

- _Skip usage of temp file when writing to S3:_
  When writing a file to S3, the file system currently writes a temp file locally and syncs at the end by creating a file on S3. This coupled with the fact that S3 provides atomic uploads of files means that we will never see incomplete data uploaded for a file. In this scenario, the current usage of temp files when writing anything to the s3 file system (such as for Checkpoints, SavedModels) are unnecessary. I've removed such usage of temp files for S3 as they not just are unnecessary but cause issues with S3's consistency model. (It may so happen that after writing to a temp file, the file is not visible while moving to final location because changes haven't propagated yet. )

- _Skip usage of temp directory when saving Checkpoints and SavedModels to S3:_ 
  When saving Checkpoints and SavedModels, there is also the use of a temp directory where all files are uploaded and then moved to the final location. This is even more problematic given S3's eventual consistency model. This sequence of operations involves the writing to temp location, listing of files in temp location, copying each file from this temp location to final location, and deleting the temp location. There are multiple issues users have experienced because of this sequence. When listing files in temp directory, in some cases not all files show up in the list. This causes some files to not be copied. While deleting directory, there is another listing of all files in the directory. Again this list may be inconsistent and the deleting of directory fails. So the usage of such temp directory has been bypassed for S3 file system.

- The above two changes have been implemented by introducing a NeedsTempLocation method for the filesystem and exposing it to python through gfile similar to how IsDirectory works. This method needs to be used multiple places both in python and CPP. By default for any file system other than S3, it returns False, which allows us to skip the usage of temp locations.

- The above issues for Checkpointing and SavedModels have been addressed and verified for both code paths of Estimator based training, Session based training as well as Keras. 

- _Retry failed operations before crashing:_
  It is recommended to use retries to handle exceptions in the case of unexpected errors due to the eventual consistency paradigm. Moved the class retrying_file_system from `platform/cloud` to `platform/` and reused that for S3 file system.

**Reliability and Performance**
- _Use TransferManager from the AWS CPP SDK to upload files to S3_ 
Using PUT for uploads as is done currently has a few limitations. It doesn't allow the upload of files larger than 5GB which in our experience, users have run into. It also doesn't retry in the case of crashes. Using the TransferManager allows us to upload larger files, upload files faster by using multi-part upload, and retry only the failed parts in case of crashes.

  | Time to upload 2GB | Time to upload 4GB | Time to upload 5.5GB
-- | -- | -- | --
Official Tensorflow r1.12 | 103 sec | 116 sec | Crash: Entity too large
Tensorflow 1.12 with S3 patch | 9 sec | 20 sec | 28 sec
Speedup | 11x | 6x |  

I request the maintainers to review, and provide feedback.

### CODE REVIEW

**Comment by qlzh727** [comment]
Can u please rebase and resolve the merge conflict? Thanks

**Comment by rahul003** [comment]
Sure, am on it. Looks like there was a change to the v2 API two days back. Integrating that.

**Comment by rahul003** [comment]
I rebased the changes onto master. Please check.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @yongtang, @k-w-w: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

File: tensorflow/core/platform/s3/s3_file_system.cc [hunk 1]
```diff
@@ -176,6 +193,7 @@ class S3RandomAccessFile : public RandomAccessFile {
 
   Status Read(uint64 offset, size_t n, StringPiece* result,
               char* scratch) const override {
+    VLOG(1) << "ReadFilefromS3 s3://" << bucket_ << "/" << object_;
     Aws::S3::Model::GetObjectRequest getObjectRequest;
```
Comment (Jeffwan) [inline tensorflow/core/platform/s3/s3_file_system.cc]: TransferManager not only helps on the upload part but also downloading part. I think migrate GetObject low level api to transfer manager will be really helpful.

Comment (Jeffwan) [reply tensorflow/core/platform/s3/s3_file_system.cc]: @rahul003  Do you have plan that also add this support?

Comment (rahul003) [reply tensorflow/core/platform/s3/s3_file_system.cc]: Hi jeffwan@, TransferManager indeed helps with the download part as well, but the file system interface seems to require implementing the ability to read a file from a given offset for a specific length. TransferManager only supports fetching the full file. It would require some more work. This can perhaps be added in the future.

**Comment by rthadur** [comment]
@rahul003 can you please resolve conflicts

**Comment by rthadur** [comment]
I'm going to go ahead and close this PR, because it seems to have stalled. If you're still interested in pursing this (and responding to my comments), please feel free to reopen!
